### PR TITLE
LIR: Fix panic. Enable proptest for `ReducePlan`

### LINKED ISSUE #11928: Add protobuf support for `mz_dataflow_types::plan::reduce::ReducePlan`

### PR DESCRIPTION
**Pull request by wangandi**
### Motivation

  * This PR fixes a previously unreported bug.

Materialize crashes if you have the query hint `option(expected_group_size=<number greater than usize::MAX/16>)`.

Repro commands:
```
create table t1(f1 int, f2 int);
select max(f1) from t1 option(expected_group_size=111311797981107522680011);
```

The reason is that we have a loop that continues multiplies a number by until it is equal than or greater than `expected_group_size`. But if `expected_group_size` is greater than `usize::MAX/16`, then doing this multiplication loop will result in integer overflow.

Fixed by capping expected_group_size to be less than or equal to `usize::MAX/16`.

  * This PR adds a known-desirable feature.

Closes #11928.

Previously, proptest for `ReducePlan` would crash due to stack overflow. This gets rid of the stack overflow by:
* limiting the number of aggregations that we add to a single `ReducePlan`
* generating `ReducePlan`s by supplying arguments to `ReducePlan::create_from` rather than making `proptest` construct arbitrary sequences of `AccumulablePlan`, `HierarchicalPlan`, etc.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

* Fixes a panic that would occur if the query hint `expected_group_size` were set to a value greater than `usize::MAX/16`.

### CODE REVIEW

File: src/dataflow-types/src/plan/reduce.rs [hunk 1]
```diff
@@ -164,6 +163,35 @@ pub enum ReducePlan {
     Collation(CollationPlan),
 }
 
+impl Arbitrary for ReducePlan {
```
Comment (frankmcsherry) [inline src/dataflow-types/src/plan/reduce.rs:166]: What is `Arbitrary` and can we get an explainer here?

File: src/dataflow-types/src/plan/reduce.rs [hunk 2]
```diff
@@ -739,7 +771,6 @@ impl ReducePlan {
                         buckets.push(current as u64);
                         current *= 16;
```
Comment (frankmcsherry) [inline src/dataflow-types/src/plan/reduce.rs:772]: It seems like the fix should be to make this `saturating_mul`?

**Review by frankmcsherry — COMMENTED** [review]
I might be confused, but it seems like a simpler fix is to correct the instance of overflow to one that saturates instead. Or bail out of the loop before performing the multiplication. Tweaking the arguments as done here seem to make it more likely to change the code in one place and leave crashes

File: src/dataflow-types/src/plan/reduce.rs (hunk 2 above)
Comment (wangandi) [reply src/dataflow-types/src/plan/reduce.rs:772]: Indeed it should be! I just forgot on Friday that `saturating_mul` existed. Thanks for pointing this out!

File: src/dataflow-types/src/plan/reduce.rs (hunk 1 above)
Comment (wangandi) [reply src/dataflow-types/src/plan/reduce.rs:173]: The `Arbitrary` trait should document itself (see example here:
https://dev.materialize.com/api/rust-private/mz_expr/scalar/func/enum.BinaryFunc.html#impl-Arbitrary), but basically it generates arbitrary instances of the type in order to test some property of it. We have been using it in the protobuf support work to ensure that an arbitrary compute command can be translated to and back from protobuf without loss of information.

I have added a comment though, on this particular implementation for `ReducePlan`.

**Review by frankmcsherry — APPROVED** [review]
Works for me, except there seem to be two unrelated changes going on: fixing a crash, and changing the `Arbitrary` implementation.

**Comment by wangandi** [comment]
> Works for me, except there seem to be two unrelated changes going on: fixing a crash, and changing the Arbitrary implementation.

I think of them as semi-related because the `Arbitrary` implementation change means that our unit tests verify that the crash has been fixed. The `proptest` will call `ReducePlan::create_from` with a bunch of sets of randomly generated values, and approximately 5% of those sets would have triggered the crash.

**Review by lluki — APPROVED** [review]

File: src/dataflow-types/src/plan/reduce.rs [hunk 3]
```diff
@@ -164,6 +163,35 @@ pub enum ReducePlan {
...
+        (
+            proptest::collection::vec(any::<AggregateExpr>(), 0..8),
+            any::<bool>(),
+            any::<bool>(),
+            (0..u32::MAX),
+        )
+            .prop_map(|(exprs, monotonic, any_expected_size, size)| {
+                let expected_group_size = if any_expected_size {
+                    if size > 4_000_000_000 {
+                        // Test a very large edge case approximately 5% of the time.
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/dataflow-types/src/plan/reduce.rs]: I decided to try for a logarithmic distribution instead. Let me know if whether you think the logarithmic distribution is idiomatic.

[the remark this answers is not in the document]
Comment (lluki) [reply src/dataflow-types/src/plan/reduce.rs]: Looks good!

**Pull request merged**